### PR TITLE
Aut-3046: Remove user lockout from requesting too many sms code

### PR DESCRIPTION
**Pull request by di-fabs**
## What
- Remove user's lockout action from requesting too many sms code

## Checklist

- [ ] Impact on orch and auth mutual dependencies has been checked.

- [ ] A UCD review has been performed.

### CODE REVIEW

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java [hunk 1]
```diff
@@ -249,28 +249,58 @@ private Optional<ErrorResponse> validateCodeRequestAttempts(
...
-                    "User has requested too many OTP codes. Setting block with prefix: {}",
-                    newCodeRequestBlockPrefix);
-            codeStorageService.saveBlockedForEmail(
-                    email, newCodeRequestBlockPrefix, configurationService.getLockoutDuration());
-            LOG.info("Resetting code request count");
-            sessionService.save(
-                    session.resetCodeRequestCount(NotificationType.MFA_SMS, journeyType));
+            LOG.warn("User has requested too many OTP codes.");
+
+            blockReauthenticatingUserWhenReauthenticationLogOffNotSupported(
```
Comment (BeckaL) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java:254]: Could we have some tests for these bits of behaviour?

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java [hunk 2]
```diff
@@ -249,28 +249,58 @@ private Optional<ErrorResponse> validateCodeRequestAttempts(
...
+                    email, journeyType, newCodeRequestBlockPrefix);
+
+            blockUsersOnAllJourneysOtherThanReauthenticatingUsers(
```
Comment (BeckaL) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java:257]: Like these nice names, very nice to read :-)

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/SendNotificationHandler.java [hunk 3]
```diff
@@ -369,8 +369,12 @@ private Optional<ErrorResponse> isCodeRequestAttemptValid(
             LOG.info(
                     "User has requested too many OTP codes. Setting block with prefix: {}",
                     newCodeRequestBlockPrefix);
-            codeStorageService.saveBlockedForEmail(
-                    email, newCodeRequestBlockPrefix, configurationService.getLockoutDuration());
+            if (!configurationService.supportReauthSignoutEnabled()) {
```
Comment (BeckaL) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/SendNotificationHandler.java:372]: Is this the complete condition? I.e. are we always in reauth journeys at this point? If not, does this not mean when the flag is on we'd not be blocking for all journeys?

Also, if we could have some tests for these scenarios that'd be good :-)

Comment (andrew-moores) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/SendNotificationHandler.java]: These changes aren't needed.  This lambda is only called from the resend-email-code-controller which is not a route for re-authenticating users.

File: frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java [hunk 4]
```diff
@@ -380,9 +383,22 @@ void shouldReturn400IfUserHasReachedTheSmsSignInCodeRequestLimit(JourneyType jou
         assertThat(result, hasJsonBody(ErrorResponse.ERROR_1025));
 
         var codeRequestType = CodeRequestType.getCodeRequestType(MFAMethodType.SMS, journeyType);
-        verify(codeStorageService)
-                .saveBlockedForEmail(
-                        EMAIL, CODE_REQUEST_BLOCKED_KEY_PREFIX + codeRequestType, LOCKOUT_DURATION);
+
+        if (journeyType == JourneyType.REAUTHENTICATION) {
```
Comment (BeckaL) [inline frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java:387]: Suggestion:

```        
if (journeyType != JourneyType.REAUTHENTICATION || !reauthEnabled) {
                verify(codeStorageService)
                        .saveBlockedForEmail(
                                EMAIL,
                                CODE_REQUEST_BLOCKED_KEY_PREFIX + codeRequestType,
                                LOCKOUT_DURATION);
        } else {
            verifyNoInteractions(codeStorageService);
        }

verify(sessionService).save(argThat((session) ->
                session.getCodeRequestCount(NotificationType.MFA_SMS, journeyType) == 0));
```

(I found the nested conditions a bit confusing, and would be good to verify what happens on a reauth journey with reauth enabled, and to clarify that we wipe the count in any case)

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java [hunk 5]
```diff
@@ -249,28 +249,58 @@ private Optional<ErrorResponse> validateCodeRequestAttempts(
...
             return Optional.of(ErrorResponse.ERROR_1027);
         }
+
         return Optional.empty();
     }
+
+    private void blockReauthenticatingUserWhenReauthenticationLogOffNotSupported(
+            String email, JourneyType journeyType, String newCodeRequestBlockPrefix) {
+        if (journeyType == JourneyType.REAUTHENTICATION
+                && !configurationService.supportReauthSignoutEnabled()) {
```
Comment (BeckaL) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java:285]: Not strictly related to this PR, but I notice we have duplicate configuration service methods reading from the same env var (`supportReauthSignoutEnabled` and `isReauthSignoutEnabled`) - I think we should raise a subsequent PR after this one to consolidate

**Review by BeckaL — COMMENTED** [review]
Generally looks good to me. One suggestion about test code, and a note about a subsequent small PR I think we should raise.

File: frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java (hunk 4 above)
Comment (andrew-moores) [reply frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java]: Added some clarifying methods to make this clearer.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java (hunk 5 above)
Comment (andrew-moores) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandler.java:285]: Removed isReauthSignoutEnabled in favour of supportReauthSignoutEnabled.

File: frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java (hunk 4 above)
Comment (BeckaL) [reply frontend-api/src/test/java/uk/gov/di/authentication/frontendapi/lambda/MfaHandlerTest.java]: Cool LGTM, I think it's still worth that extra assertion ```verify(sessionService).save(argThat((session) ->
                session.getCodeRequestCount(NotificationType.MFA_SMS, journeyType) == 0))``` being added in

**Review by BeckaL — APPROVED** [review]
LGTM!

**Pull request merged**